From 47da4061a3f9844069c054640f15a3cc04a82b22 Mon Sep 17 00:00:00 2001
From: Emil Renner Berthing <kernel@esmil.dk>
Date: Sun, 29 Mar 2015 20:03:53 +0200
Subject: [PATCH] ARM: dts: imx51: add SAHARA v4 entry

Signed-off-by: Emil Renner Berthing <kernel@esmil.dk>
---
 arch/arm/boot/dts/imx51.dtsi | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/arch/arm/boot/dts/imx51.dtsi b/arch/arm/boot/dts/imx51.dtsi
index c0116cf..e951f9a 100644
--- a/arch/arm/boot/dts/imx51.dtsi
+++ b/arch/arm/boot/dts/imx51.dtsi
@@ -247,6 +247,15 @@
 					bus-width = <4>;
 					status = "disabled";
 				};
+
+				sahara: crypto@83ff8000 {
+					compatible = "fsl,imx53-sahara";
+					reg = <0x83ff8000 0x4000>;
+					interrupts = <19 20>;
+					clocks = <&clks IMX5_CLK_SAHARA_IPG_GATE>,
+					         <&clks IMX5_CLK_SAHARA_IPG_GATE>;
+					clock-names = "ipg", "ahb";
+				};
 			};
 
 			usbotg: usb@73f80000 {
