--- a/drivers/ata/ahci_sunxi.c	2015-08-09 11:13:34.922768480 +0200
+++ b/drivers/ata/ahci_sunxi.c	2015-08-09 11:14:20.386766709 +0200
@@ -29,12 +29,6 @@
 
 #define DRV_NAME "ahci-sunxi"
 
-/* Insmod parameters */
-static bool enable_pmp;
-module_param(enable_pmp, bool, 0);
-MODULE_PARM_DESC(enable_pmp,
-	"Enable support for sata port multipliers, only use if you use a pmp!");
-
 #define AHCI_BISTAFR	0x00a0
 #define AHCI_BISTCR	0x00a4
 #define AHCI_BISTFCTR	0x00a8
@@ -198,14 +192,6 @@ static int ahci_sunxi_probe(struct platf
 	hpriv->flags = AHCI_HFLAG_32BIT_ONLY | AHCI_HFLAG_NO_MSI |
 		       AHCI_HFLAG_YES_NCQ;
 
-	/*
-	 * The sunxi sata controller seems to be unable to successfully do a
-	 * soft reset if no pmp is attached, so disable pmp use unless
-	 * requested, otherwise directly attached disks do not work.
-	 */
-	if (!enable_pmp)
-		hpriv->flags |= AHCI_HFLAG_NO_PMP;
-
 	rc = ahci_platform_init_host(pdev, hpriv, &ahci_sunxi_port_info,
 				     &ahci_platform_sht);
 	if (rc)
